device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -0,0 +1,456 @@
+                break;
+            default:
+                throw new IllegalArgumentException("Multiplexing is only supported for AMQPS and AMQPS_WS");
+        }
+
+        this.deviceClientList = new ArrayList<>();
+        this.hostName = hostName;
+        this.protocol = protocol;
+        this.proxySettings = options != null ? options.getProxySettings() : null;
+        long sendPeriod = options != null ? options.getSendPeriod() : DEFAULT_SEND_PERIOD_MILLIS;
+        long receivePeriod = options != null ? options.getReceivePeriod() : DEFAULT_RECEIVE_PERIOD_MILLIS;
+
+        if (sendPeriod < 0)
+        {
+            throw new IllegalArgumentException("send period can not be negative");
+        }
+        else if (sendPeriod == 0) //default builder value for this option, signals that user didn't set a value
+        {
+            sendPeriod = DEFAULT_SEND_PERIOD_MILLIS;
+        }
+
+        if (receivePeriod < 0)
+        {
+            throw new IllegalArgumentException("receive period can not be negative");
+        }
+        else if (receivePeriod == 0) //default builder value for this option, signals that user didn't set a value
+        {
+            receivePeriod = DEFAULT_RECEIVE_PERIOD_MILLIS;
+        }
+
+        this.sslContext = options != null ? options.getSslContext() : null;
+        this.deviceIO = new DeviceIO(hostName, protocol, sslContext, proxySettings, sendPeriod, receivePeriod);
+    }
+
+    /**
+     * Opens this multiplexing client. This may be done before or after registering any number of device clients.
+     * <p>
+     * This call behaves synchronously, so if it returns without throwing, then all registered device clients were
+     * successfully opened.
+     * <p>
+     * If this client is already open, then this method will do nothing.
+     * <p>
+     * @throws IOException If any IO errors occur while opening the multiplexed connection.
+     */
+    public void open() throws IOException
+    {
+        synchronized (this.operationLock)
+        {
+            log.info("Opening multiplexing client");
+            this.deviceIO.open();
+            log.info("Successfully opened multiplexing client");
+        }
+    }
+
+    /**
+     * Close this multiplexing client. This will close all active device sessions as well as the AMQP connection.
+     * <p>
+     * If this client is already closed, then this method will do nothing.
+     * <p>
+     * Once closed, this client can be re-opened. It will preserve all previously registered device clients.
+     * <p>
+     * @throws IOException If any exception occurs while closing the connection.
+     */
+    public void close() throws IOException
+    {
+        synchronized (this.operationLock)
+        {
+            log.info("Closing multiplexing client");
+            for (DeviceClient deviceClient : this.deviceClientList)
+            {
+                deviceClient.closeFileUpload();
+            }
+
+            if (this.deviceIO != null)
```

Comment:
q - would `deviceIO` ever be null? Would it be null if I were to call multiplexedClient.close() twice, in succession?

Reply:
It will never be null. This check was a remnant of back when the deviceIO was only created after registering the first device client. Now that the constructor builds this, and it is declared final, I can remove this null check

Reply:
Closing/opening the client has no affect on what deviceIO object is set here. Only the constructor sets it. No one else sets it to null